Publish repo to Docker on CI

fedgiac opened the pull request:
Adds a script that publishes the entire repo to Docker on every push to main and new tag.

The code in the PR is almost entirely copied from [gnosis/gp-v2-trading-bot](https://github.com/gnosis/gp-v2-trading-bot).

I'll ask devops for the needed secrets and add them to the services before merging.

The Docker image will be used by out withdraw service, and must support executing arbitrary hardhat scripts.

### Test plan

Note that I still have no Docker credentials, so the output of this test plan is only useful to check that the result of running `yarn install --frozen` is the same as running it locally.
<details><summary>bash src/docker/deploy.sh test</summary>

```
$  export DOCKERHUB_PROJECT=gp-v2-contracts
$ bash src/docker/deploy.sh test
Trying to pull docker.io/gnosispm/gp-v2-contracts:test...
Error: initializing source docker://gnosispm/gp-v2-contracts:test: reading manifest test in docker.io/gnosispm/gp-v2-contracts: errors:
denied: requested access to the resource is denied
unauthorized: authentication required

STEP 1/6: FROM node:14.16.1-alpine3.13
STEP 2/6: WORKDIR /usr/src/app/
--> Using cache 99bfacf928eb2ee6744d1aa7a2892b32ec8aeca1aa329d83f64b9161cc40ef84
--> 99bfacf928e
STEP 3/6: COPY . .
--> ecffe73afd5
STEP 4/6: RUN apk add --no-cache tini git python3 linux-headers eudev-dev libusb-dev build-base
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/47) Upgrading musl (1.2.2-r0 -> 1.2.2-r1)
(2/47) Installing binutils (2.35.2-r1)
(3/47) Installing libmagic (5.39-r0)
(4/47) Installing file (5.39-r0)
(5/47) Installing libgomp (10.2.1_pre1-r3)
(6/47) Installing libatomic (10.2.1_pre1-r3)
(7/47) Installing libgphobos (10.2.1_pre1-r3)
(8/47) Installing gmp (6.2.1-r0)
(9/47) Installing isl22 (0.22-r0)
(10/47) Installing mpfr4 (4.1.0-r0)
(11/47) Installing mpc1 (1.2.0-r0)
(12/47) Installing gcc (10.2.1_pre1-r3)
(13/47) Installing musl-dev (1.2.2-r1)
(14/47) Installing libc-dev (0.7.2-r3)
(15/47) Installing g++ (10.2.1_pre1-r3)
(16/47) Installing make (4.3-r0)
(17/47) Installing fortify-headers (1.1-r0)
(18/47) Installing patch (2.7.6-r7)
(19/47) Installing build-base (0.5-r2)
(20/47) Installing udev-init-scripts (34-r0)
Executing udev-init-scripts-34-r0.post-install
(21/47) Installing eudev-libs (3.2.9-r3)
(22/47) Installing libblkid (2.36.1-r1)
(23/47) Installing xz-libs (5.2.5-r0)
(24/47) Installing zstd-libs (1.4.9-r0)
(25/47) Installing kmod-libs (28-r0)
(26/47) Installing eudev (3.2.9-r3)
(27/47) Installing pkgconf (1.7.3-r0)
(28/47) Installing eudev-dev (3.2.9-r3)
(29/47) Installing ca-certificates (20191127-r5)
(30/47) Installing brotli-libs (1.0.9-r3)
(31/47) Installing nghttp2-libs (1.42.0-r1)
(32/47) Installing libcurl (7.78.0-r0)
(33/47) Installing expat (2.2.10-r1)
(34/47) Installing pcre2 (10.36-r0)
(35/47) Installing git (2.30.2-r0)
(36/47) Installing libusb (1.0.24-r1)
(37/47) Installing libusb-dev (1.0.24-r1)
(38/47) Installing linux-headers (5.7.8-r0)
(39/47) Installing libbz2 (1.0.8-r1)
(40/47) Installing libffi (3.3-r2)
(41/47) Installing gdbm (1.19-r0)
(42/47) Installing ncurses-terminfo-base (6.2_p20210109-r0)
(43/47) Installing ncurses-libs (6.2_p20210109-r0)
(44/47) Installing readline (8.1.0-r0)
(45/47) Installing sqlite-libs (3.34.1-r0)
(46/47) Installing python3 (3.8.10-r0)
(47/47) Installing tini (0.19.0-r0)
Executing busybox-1.32.1-r6.trigger
Executing eudev-3.2.9-r3.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 266 MiB in 62 packages
--> 10da87276d2
STEP 5/6: RUN yarn install --frozen
yarn install v1.22.5
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime@7.13.10" has unmet peer dependency "@babel/core@^7.0.0-0".
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime > babel-plugin-polyfill-corejs2@0.1.10" has unmet peer dependency "@babel/core@^7.0.0-0".
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime > babel-plugin-polyfill-corejs3@0.1.7" has unmet peer dependency "@babel/core@^7.0.0-0".
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime > babel-plugin-polyfill-regenerator@0.1.6" has unmet peer dependency "@babel/core@^7.0.0-0".
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime > babel-plugin-polyfill-corejs2 > @babel/helper-define-polyfill-provider@0.1.5" has unmet peer dependency "@babel/core@^7.4.0-0".
warning "@gnosis.pm/util-contracts > @truffle/hdwallet-provider > @trufflesuite/web3-provider-engine > eth-block-tracker > @babel/plugin-transform-runtime > babel-plugin-polyfill-corejs2 > @babel/helper-define-polyfill-provider > @babel/helper-compilation-targets@7.13.13" has unmet peer dependency "@babel/core@^7.0.0".
warning "hardhat-gas-reporter > eth-gas-reporter@0.2.22" has unmet peer dependency "@codechecks/client@^0.1.0".
warning " > ts-node@10.2.1" has unmet peer dependency "@types/node@*".
[4/4] Building fresh packages...
Done in 61.27s.
--> 0be8ef690cd
STEP 6/6: ENTRYPOINT ["/sbin/tini", "--"]
COMMIT gp-v2-contracts
--> 8336415d84d
Successfully tagged localhost/gp-v2-contracts:latest
8336415d84dcd9a3870bc3d30d62d017fbc2043782cefd1f9208bf837ed86996
Getting image source signatures
Error: trying to reuse blob sha256:b2d5eeeaba3a22b9b8aa97261957974a6bd65274ebd43e1d81d0a7b8b752b116 at destination: checking whether a blob sha256:b2d5eeeaba3a22b9b8aa97261957974a6bd65274ebd43e1d81d0a7b8b752b116 exists in docker.io/gnosispm/gp-v2-contracts: errors:
denied: requested access to the resource is denied
error parsing HTTP 401 response body: unexpected end of JSON input: ""
```
</details>

fedgiac committed:
Publish repo to Docker on CI

bh2smith reviewed (approved):
Boss. Was there a reason for doing this? Like will we now fetch the image for our deployments instead of what we do now?

fedgiac committed:
Add dependencies for yarn install

fedgiac commented:
Right, the motivation is missing from the description, I added it! Basically, we can use this Docker image to run our withdraw service.

Now that I have a proper internet connection, I noticed that the test plan is actually not working as intended and that was not just an artifact of a bad internet connection. I fixed the issue (by adding the dependencies needed for `yarn install`) and successfully ran the test plan, see updated description.

fedgiac committed:
Update env variable name to actual name

fedgiac committed:
Merge branch 'main' into docker-deployment

fedgiac commented:
Coverall is [down](https://status.coveralls.io/), and since tests are a prerequisite for deploying it means we can't test if the login is working. I'll look into making the coverall step optional in another PR.

fedgiac committed:
Merge branch 'main' into docker-deployment

nlordell commented on .github/workflows/CI.yml:
```diff
@@ -35,3 +35,21 @@ jobs:
+      - name: Dockerhub login
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_NAME }}
```
ubernit: I know we use this elsewhere but
```suggestion
          username: ${{ secrets.DOCKER_USER }}
```

:stuck_out_tongue:

fedgiac replied:
Devops set the variable name and I don't know the secrets, so I had to changed the variable from `DOCKER_USER` to `DOCKER_NAME` to make it work. :slightly_smiling_face: 
Should I ask them to change it?

nlordell replied:
Nah, not worth the trouble in my opinion :stuck_out_tongue:

fedgiac committed:
Merge branch 'main' into docker-deployment

fedgiac commented:
First Docker deployment was successful! See [passing CI run](https://github.com/gnosis/gp-v2-contracts/runs/3661022123) and [Dockerhub deployment](https://hub.docker.com/layers/gnosispm/gp-v2-contracts/staging/images/sha256-ab17c620e6f76dc2d7ef275403c1a0b7624d4b671f15a46fee92754c3100010d).
